crates/biome_html_analyze/src/lint/nursery/no_duplicate_classes.rs

```diff
@@ -0,0 +1,183 @@
+    type State = DuplicateClassesState;
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Option<Self::State> {
+        let attribute = ctx.query();
+
+        // Only check "class" attribute
+        let name = attribute.name().ok()?;
+        let name_token = name.value_token().ok()?;
+        if name_token.text_trimmed() != "class" {
+            return None;
+        }
+
+        // Get the attribute value
+        let initializer = attribute.initializer()?;
+        let value = initializer.value().ok()?;
+        let html_string = value.as_html_string()?;
+        let value_token = html_string.value_token().ok()?;
+        let value_text = value_token.text_trimmed();
+
+        // Check if single-quoted
+        let is_single_quote = value_text.starts_with('\'');
+
+        // Get the inner string (without quotes)
+        let inner_text = inner_string_text(&value_token);
+
+        // Split by whitespace and track duplicates
+        let mut seen: HashSet<&str> = HashSet::default();
+        let mut duplicate_set: HashSet<&str> = HashSet::default();
+        let mut deduplicated_parts: Vec<&str> = Vec::new();
+
+        for class in inner_text.text().split_whitespace() {
+            if seen.contains(class) {
+                duplicate_set.insert(class);
+            } else {
+                seen.insert(class);
+                deduplicated_parts.push(class);
+            }
+        }
+
+        if duplicate_set.is_empty() {
+            return None;
+        }
+
+        let deduplicated = deduplicated_parts.join(" ");
+        let mut duplicates: Vec<Box<str>> = duplicate_set.into_iter().map(Into::into).collect();
+        duplicates.sort();
+
+        Some(DuplicateClassesState {
+            deduplicated: deduplicated.into(),
+            duplicates: duplicates.into_boxed_slice(),
+            is_single_quote,
+        })
+    }
+
+    fn diagnostic(ctx: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        let node = ctx.query();
+        let duplicates_str = state
+            .duplicates
+            .iter()
+            .map(|s| format!("`{}`", s))
+            .collect::<Vec<_>>()
+            .join(", ");
+
+        let message = if state.duplicates.len() > 1 {
+            format!("Duplicate CSS classes detected: {duplicates_str}")
+        } else {
+            format!("Duplicate CSS class detected: {duplicates_str}")
+        };
+
+        Some(
+            RuleDiagnostic::new(rule_category!(), node.range(), markup! {{message}}).note(markup! {
+                "Remove duplicate classes to improve readability."
+            }),
+        )
+    }
+
+    fn action(ctx: &RuleContext<Self>, state: &Self::State) -> Option<HtmlRuleAction> {
+        let mut mutation = ctx.root().begin();
+        let attribute = ctx.query();
+        let deduplicated = &state.deduplicated;
+
+        // Get the initializer and value
```

Comment:
We're repeating the same operations we're doing in the run function. Why don't we use the state to pass the node we will replace, instead?